gst-plugin-pravega/src/pravegasink/imp.rs

```diff
@@ -73,11 +80,135 @@ pub enum TimestampMode {
+    interval_seconds: u64,
+    retention_policy: RetentionPolicy,
+    factory: ClientFactory,
+    index_searcher: IndexSearcher<ByteReader>,
+    index_writer: ByteWriter,
+    data_writer: ByteWriter,
+}
+
+impl RetentionMaintainer {
+    fn new(element: super::PravegaSink, interval_seconds: u64, retention_policy: RetentionPolicy, factory: ClientFactory, index_scoped_segment: ScopedSegment, data_scoped_segment: ScopedSegment) -> Self {
+        let index_reader = factory.create_byte_reader(index_scoped_segment.clone());
+        let index_writer = factory.create_byte_writer(index_scoped_segment);
+        let data_writer = factory.create_byte_writer(data_scoped_segment);
+        let index_searcher = IndexSearcher::new(index_reader);
+        Self {
+            element,
+            interval_seconds,
+            retention_policy,
+            factory,
+            index_searcher, 
+            index_writer,
+            data_writer,
+        }
+    }
+
+    fn days_to_seconds(days: f64) -> i128 {
+        let seconds = days * 24.0 * 60.0 * 60.0;
+        seconds.round() as i128
+    }
+
+    fn run(mut self, thread_stop_rx: Receiver<()>) -> Option<JoinHandle<()>> {
+        let (seconds, bytes) = match self.retention_policy {
+            RetentionPolicy::Days(days) => (Some(RetentionMaintainer::days_to_seconds(days)), None),
+            RetentionPolicy::Bytes(bytes) => (None, Some(bytes)),
+            RetentionPolicy::DaysAndBytes(days, bytes) => (Some(RetentionMaintainer::days_to_seconds(days)), Some(bytes)),
+            _ => (None, None),
+        };
+
+        if seconds.is_none() && bytes.is_none() {
+            return None;
+        }
+        
+        gst_info!(CAT, obj: &self.element, "start: retention_maintainer_interval_seconds={}", self.interval_seconds);
+        let handle = thread::spawn(move || {
+            loop {
+                if let Some(sec) = seconds {
+                    let truncate_at_timestamp = PravegaTimestamp::now() - sec * SECOND;
+                    gst_info!(CAT, obj: &self.element, "Truncating prior to {}", truncate_at_timestamp);
+
+                    let search_result = self.index_searcher.search_timestamp_and_return_index_offset(truncate_at_timestamp, SearchMethod::Before);
+                    if let Ok(result) = search_result {
+                        let runtime = self.factory.runtime();
+                        runtime.block_on(self.index_writer.truncate_data_before(result.1 as i64)).unwrap();
+                        gst_info!(CAT, obj: &self.element, "Index truncated at offset {}", result.1);
+                        runtime.block_on(self.data_writer.truncate_data_before(result.0.offset as i64)).unwrap();
+                        gst_info!(CAT, obj: &self.element, "Data truncated at offset {}", result.0.offset);
+                    }
+                }
+
+                // break the loop to stop the thread
+                match thread_stop_rx.try_recv() {
```

Comment:
The problem here is that it will only check this every 15 minutes because of the sleep. Instead, use ` recv_timeout` with a timeout of 15 minutes.

Reply:
good point.